### ISSUE TITLE
 Lab2 Review by Davide Sferrazza s326619

### ISSUE THREAD

**Issue by FarInHeight** [Gabbo62/ComputationalIntelligence#1]
## Intro

Hello 👋🏻, Gabriele!

First of all, I'm glad that you finished the 2nd lab in time. Well done 💪🏻!

I'd like to divide my review into two parts:
- the first part analyzes your `fixed_rule` function;
- the second part analyzes your evolved rule implementation.

## `fixed_rule` function

I appreciate your idea of implementing a function that leaves a value of `sum(state.rows)` equal to a power of 2 reduced by 1. If I understand it correctly, it should work.

The result of 81.2% win rate seems great.

However, looking at your implementation, I see that you generate all the `spicy_moves` by checking also if:
```python
sum(state.rows) != 0
```
The check for `sum(state.rows) != 0` (`ns != 0` in your code) is not needed, because in the case in which `sum(state.rows)` is 0, the `fixed_rule` function won't be called, because the game is over. \
You can try to confirm this by playing several games with this extra condition removed.

## Evolved rule implementation

Your implementation is pretty clear, although I don't think it's a correct evolution strategy (assuming I understood your intentions). \
What you do is that for each state that a Nim game could be in, you create an offspring from the current state which is made up of moves, and decide the best move to apply using the nim-sum value as fitness. The smaller it is, the better the move. \
In this way the only real valued parameter you are training is the `tweak_var` variable.

I expected you to create an agent capable of playing alone, without the need to generate an offspring just to play a single move.

All in all, this solution is quite surprising to me, because I never thought to train something like you did. Your idea is really innovative for me. I am glad to have known this alternative way of thinking to my own. 

## Final observations

I encourage you to extend your work to play also in subtraction game mode, i.e. by limiting each agent to take at most $k$ objects from each row.

Ultimately, good work 🙂!
I hope you'll find this review useful and good luck for the next labs 🍀😉!

**Comment by Gabbo62** [comment]
Hello Davide,  
I want to thank you for your useful analysis on my code.

I want to let you know that in this part you probably misunderstood the meaning of ns in the generation of spicy_moves.
> The check for sum(state.rows) != 0 (ns != 0 in your code) is not needed, because in the case in which sum(state.rows) is 0, the fixed_rule function won't be called, because the game is over. 

In fact the value of ns is computed inside the `analize` function (I just noticed the typo), as you understood correctly, with the sum of the remaining sticks in the game.  
The misunderstanding comes here because the computation of the remaining sticks is done on the _hypothetically_ obtained state after considering the generated move.
```
tmp = deepcopy(raw)
tmp.nimming(ply)  
cooked["possible_moves"][ply] = remain_sum(tmp)
```
_Where raw is the actual game state and remain_sum is the sum of the sticks in this case called on the state obtained after the move is applied_

So it's correct what you said (that the function is not called if the sum of the _real_ game is 0) but the meaning of this sum is connected to the future state and not the present one.  
By doing that i'm excluding from the useful moves the ones that bring the player to a direct lost (because otherwise `np.log2(ns+1) % 1 == 0` will be true and the move will not be discarded).

For what is concerned on the second part I think that you understood correctly my intentions.
> What you do is that for each state that a Nim game could be in, ...

Just to be sure that we are thinking the same thing about this statement i prefer to add a little comment: the generation of the offspring is done in response to a move made by the other agent and not "for each state" we could find us in.

Said that, I didn't thought I needed to create an agent able to play alone but that he just needed to respond to what it is presented to him. That is an interesting new way to think about the problem.   
I can try to implement a new solution based on what you told me and also add the part that consider the limitation of k element taken at most from each row.

In conclusion, I want to thank you for your very useful comment and I hope you'd like to come again to get a look on the future labs we will work on.🙂

**Comment by FarInHeight** [comment]
Hi 👋🏻.

I didn't notice what you said about your 'analyze' function. I retract what I said and replace it with "Really good work!!"😆.

With reference to
> Just to be sure that we are thinking the same thing about this statement i prefer to add a little comment: the generation of the offspring is done in response to a move made by the other agent and not "for each state" we could find us in.

This is exactly what I meant. Sorry if I wasn't clear enough 😉.

Well, good luck with your future labs, and I'll be happy to come back here to check out your future works 🙃.
